### PR TITLE
[SandboxIR] More boilerplate: Function, Argument, Constant, Instruction, OpaqueInst

### PR DESCRIPTION
**Pull request by vporpo**
A very basic implementation of sandboxir::
`Fuction`
`Argument`
`Constant`
`Instruction`
`OpaqueInst`

### CODE REVIEW

File: llvm/lib/SandboxIR/SandboxIR.cpp [hunk 1]
```diff
@@ -59,8 +59,95 @@ void sandboxir::Value::printAsOperandCommon(raw_ostream &OS) const {
     OS << "NULL ";
 }
 
+void sandboxir::Argument::printAsOperand(raw_ostream &OS) const {
```
Comment (tschuett) [inline llvm/lib/SandboxIR/SandboxIR.cpp:62]: Do you need to prefix everything with `sandboxir::`? I hoped the `using namespace sandboxir;` was sufficient.

Comment (vporpo) [reply llvm/lib/SandboxIR/SandboxIR.cpp:62]: Yeah these can be dropped, let me remove `using namespace llvm` and replace it with `using namespace llvm::sandboxir`.

File: llvm/lib/SandboxIR/SandboxIR.cpp [hunk 2]
```diff
@@ -10,28 +10,27 @@
 #include "llvm/Support/Debug.h"
 #include <sstream>
 
-using namespace llvm;
-using namespace sandboxir;
+using namespace llvm::sandboxir;
 
-sandboxir::Value::Value(ClassID SubclassID, llvm::Value *Val, Context &Ctx)
```
Comment (aeubanks) [inline llvm/lib/SandboxIR/SandboxIR.cpp:16]: please land the namespace cleanups as a separate commit (no need for review)

File: llvm/include/llvm/SandboxIR/SandboxIR.h [hunk 3]
```diff
@@ -142,16 +172,137 @@ class User : public Value {
...
+public:
+  static const char *getOpcodeName(Opcode Opc);
+#ifndef NDEBUG
+  friend raw_ostream &operator<<(raw_ostream &OS, Opcode Opc) {
+    OS << getOpcodeName(Opc);
+    return OS;
+  }
+#endif
+  /// For isa/dyn_cast.
+  static bool classof(const sandboxir::Value *From);
```
Comment (aeubanks) [inline llvm/include/llvm/SandboxIR/SandboxIR.h:236]: no definition?

File: llvm/unittests/SandboxIR/SandboxIRTest.cpp [hunk 4]
```diff
@@ -40,3 +40,22 @@ define void @foo(i32 %v1) {
...
+  ret void
+}
+)IR");
+  llvm::Function *LLVMF = &*M->getFunction("foo");
+  llvm::BasicBlock *LLVMBB = &*LLVMF->begin();
+  llvm::Instruction *LLVMAdd = &*LLVMBB->begin();
+  auto *LLVMC = cast<llvm::Constant>(LLVMAdd->getOperand(1));
+  sandboxir::Context Ctx(C);
+  auto *LLVMArg0 = LLVMF->getArg(0);
+  [[maybe_unused]] sandboxir::Function F(LLVMF, Ctx);
```
Comment (aeubanks) [inline llvm/unittests/SandboxIR/SandboxIRTest.cpp:57]: can you add some checks like `isa<Instruction/User>()`?

File: llvm/include/llvm/SandboxIR/SandboxIR.h [hunk 5]
```diff
@@ -142,16 +172,137 @@ class User : public Value {
...
 #endif
 };
 
 class Context {
 protected:
   LLVMContext &LLVMCtx;
+  /// Maps LLVM Value to the corresponding sandboxir::Value. Owns all
+  /// SandboxIR objects.
+  DenseMap<llvm::Value *, std::unique_ptr<sandboxir::Value>>
+      LLVMValueToSBValueMap;
```
Comment (tschuett) [inline llvm/include/llvm/SandboxIR/SandboxIR.h:280]: Is the `SB` needed`

Comment (vporpo) [reply llvm/include/llvm/SandboxIR/SandboxIR.h]: I removed it.

File: llvm/include/llvm/SandboxIR/SandboxIR.h (hunk 3 above)
Comment (vporpo) [reply llvm/include/llvm/SandboxIR/SandboxIR.h:236]: I added the missing definitions and added tests that exercise them.

File: llvm/lib/SandboxIR/SandboxIR.cpp [hunk 6]
```diff
@@ -58,8 +58,127 @@ void Value::printAsOperandCommon(raw_ostream &OS) const {
...
 #endif // NDEBUG
+
+const char *Instruction::getOpcodeName(Opcode Opc) {
+  switch (Opc) {
+#define DEF_VALUE(ID, CLASS)
+#define DEF_USER(ID, CLASS)
+#define OP(OPC)                                                                \
+  case Opcode::OPC:                                                            \
+    return #OPC;
+#define OPCODES(...) __VA_ARGS__
```
[the remark this answers is not in the document]
Comment (vporpo) [reply llvm/lib/SandboxIR/SandboxIR.cpp:102]: Oops it's not used yet, I will remove it.

File: llvm/include/llvm/SandboxIR/SandboxIR.h [hunk 7]
```diff
@@ -142,16 +172,137 @@ class User : public Value {
...
+};
+
+/// A sandboxir::User with operands and opcode.
+class Instruction : public sandboxir::User {
+public:
+  enum class Opcode {
+#define DEF_VALUE(ID, CLASS)
+#define DEF_USER(ID, CLASS)
+#define OP(OPC) OPC,
+#define OPCODES(...) __VA_ARGS__
```
Comment (aeubanks) [inline llvm/include/llvm/SandboxIR/SandboxIR.h:215]: unused?

File: llvm/lib/SandboxIR/SandboxIR.cpp [hunk 8]
```diff
@@ -58,8 +58,122 @@ void Value::printAsOperandCommon(raw_ostream &OS) const {
...
+    if (SBArg == nullptr)
+      OS << "NULL";
+    else
+      SBArg->printAsOperand(OS);
+    if (Idx + 1 < NumArgs)
+      OS << ", ";
+  }
+  OS << ")";
+}
+void Function::dump(raw_ostream &OS) const {
```
Comment (aeubanks) [inline llvm/lib/SandboxIR/SandboxIR.cpp:163]: it would be nice to have tests for some of these dump functions since they're non-trivial, but that can be in a followup PR

**Review by aeubanks — APPROVED** [review]
lgtm with comment addressed

File: llvm/include/llvm/SandboxIR/SandboxIR.h (hunk 7 above)
Comment (vporpo) [reply llvm/include/llvm/SandboxIR/SandboxIR.h:215]: Yeah OPCODES is not used yet. I will remove it.

File: llvm/lib/SandboxIR/SandboxIR.cpp (hunk 8 above)
Comment (vporpo) [reply llvm/lib/SandboxIR/SandboxIR.cpp:163]: OK I will test them in a followup path.
